# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: editor/libeditor/HTMLEditSubActionHandler.cpp
# Commit: 9b15baab6f77
# Full Hash: 9b15baab6f77d09eca2956276f66fee97555e4a1
# Author: Masayuki Nakano <masayuki@d-toybox.com>
# Date: 2022-09-22 09:56:47
# Regressor Bug: 1789967
# File Overlap Count: 1
# Description:
#   Bug 1789967 - part 3: Make `HTMLEditor::CollapseSelectionToEndOfLastLeafNodeOfDocument` and `HTMLEditor::InitEditorContentAndSelection` do nothing if the document is partially editable r=m_kato
#   
#   They and their callees work with the result of `GetRoot()` which is the document
#   element or the body element.  If the body is not editable, `Selection` should
#   not be updated in non-editable region nor `<br>` elements should not be
# ==============================================================================

diff -r 4d1e93c629da -r 9b15baab6f77 editor/libeditor/HTMLEditSubActionHandler.cpp
--- a/editor/libeditor/HTMLEditSubActionHandler.cpp	Thu Sep 22 06:24:38 2022 +0000
+++ b/editor/libeditor/HTMLEditSubActionHandler.cpp	Thu Sep 22 06:27:37 2022 +0000
@@ -131,6 +131,12 @@
 nsresult HTMLEditor::InitEditorContentAndSelection() {
   MOZ_ASSERT(IsEditActionDataAvailable());
 
+  // We should do nothing with the result of GetRoot() if only a part of the
+  // document is editable.
+  if (!EntireDocumentIsEditable()) {
+    return NS_OK;
+  }
+
   nsresult rv = MaybeCreatePaddingBRElementForEmptyEditor();
   if (NS_FAILED(rv)) {
     NS_WARNING(
@@ -141,8 +147,8 @@
   // If the selection hasn't been set up yet, set it up collapsed to the end of
   // our editable content.
   // XXX I think that this shouldn't do it in `HTMLEditor` because it maybe
-  //     removed by the web app and if they call `Selection::AddRange()`,
-  //     it may cause multiple selection ranges.
+  //     removed by the web app and if they call `Selection::AddRange()` without
+  //     checking the range count, it may cause multiple selection ranges.
   if (!SelectionRef().RangeCount()) {
     nsresult rv = CollapseSelectionToEndOfLastLeafNodeOfDocument();
     if (NS_FAILED(rv)) {
@@ -154,6 +160,8 @@
   }
 
   if (IsInPlaintextMode()) {
+    // XXX Should we do this in HTMLEditor?  It's odd to guarantee that last
+    //     empty line is visible only when it's in the plain text mode.
     nsresult rv = EnsurePaddingBRElementInMultilineEditor();
     if (NS_FAILED(rv)) {
       NS_WARNING(